WalletWasabi/Blockchain/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -311,11 +311,17 @@ private ProcessedResult ProcessNoLock(SmartTransaction tx)
 		&& !weAreAmongTheSender // we are not one of the senders (it is not a self-spending tx or coinjoin)
 		&& Coins.Any(c => c.HdPubKey == hdPubKey); // the destination address has already been used (address reuse)
 
-	private static void SaveInternalKeysLatestSpendingHeight(Height txHeight, IEnumerable<HdPubKey> internalKeys)
+	private static void SaveInternalKeysLatestSpendingHeight(Height txHeight, ImmutableList<SmartCoin> myInputs)
 	{
-		foreach (var spenderKey in internalKeys)
+		IEnumerable<IGrouping<HdPubKey, SmartCoin>> grouping = myInputs
+			.Where(c => c.HdPubKey.IsInternal) // Only internal ones.
+			.GroupBy(x => x.HdPubKey);
```

Comment:
Can you merge conditions here somehow 🤔  to have less iteration in the foreach by tackling on group.Any(x => !x.IsSpent())